File: sha9001.rb
require "digest"
def sha9001(data); d=data; 9001.times { d=Digest::SHA1.digest(d) }; d; end
def sanitize(v); raise ArgumentError,"ROT input is null" if v.nil?; raise ArgumentError,"ROT input contains a NUL character" if v.include?("\0"); v; end
def rotn(v,n,key); v=sanitize(v); raise TypeError,"ROT distance must be an integer" unless n.is_a?(Integer); raise TypeError,"ROT key must be bytes" unless key.is_a?(String); d=n%26; v.chars.map{|c|o=c.ord;base=o.between?(65,90)?65:o.between?(97,122)?97:nil;base ? (base+(o-base+d)%26).chr : c}.join; end
def rot13(v,k);rotn(v,13,k);end
def ebg13(v,k);rotn(v,-13,k);end

File: sha9001.py
import hashlib, hmac

def sha9001_bytes(data):
    digest = hashlib.sha1(data).digest()
    for _ in range(9000):
        digest = hashlib.sha1(digest).digest()
    return digest

def sha9001_file(path):
    with open(path, 'rb') as handle:
        return sha9001_bytes(handle.read())

def _sign(key, domain, payload): return hmac.new(key, (domain+'\0'+payload).encode(), hashlib.sha256).digest()
def _ct_equal(a,b): return hmac.compare_digest(a,b)
def _sanitize(value):
    if value is None: raise ValueError('ROT input is null')
    if '\0' in value: raise ValueError('ROT input contains a NUL character')
    return value
def _rot(value, distance):
    d=distance%26; out=[]
    for c in value:
        o=ord(c); base=65 if 65<=o<=90 else 97 if 97<=o<=122 else None
        out.append(chr(base+(o-base+d)%26) if base else c)
    return ''.join(out)
def rotn(value, distance, key):
    value=_sanitize(value); ss=_sign(key,'ROT-STRING',value); si=_sign(key,'ROT-INTEGER',str(distance))
    if not _ct_equal(ss,_sign(key,'ROT-STRING',value)) or not _ct_equal(si,_sign(key,'ROT-INTEGER',str(distance))): raise ValueError('ROTN rejected signed input')
    return _rot(value,distance)
def rot13(value,key): return rotn(value,13,key)
def ebg13(value,key): return rotn(value,-13,key)
